### DISCUSSION TITLE
[3 Novembre 2022] Pipeline

### DISCUSSION THREAD

**Discussion by Elia-Belli** [sapienzastudentsnetwork/architettura-degli-elaboratori#78] [Pipeline (M-Z)]
Il programma qui di seguito somma i numeri negativi nella sequenza fornita.
Si supponga che tutte le istruzioni impiegate fanno parte del set supportato dalla CPU in figura, ossia non si fa uso di alcuna pseudoistruzione.
Si indichino (ignorando hazard che possano concernere la syscall):
$1)$ le istruzioni tra le quali sono presenti **data hazard** – indicare i numeri di istruzione N ed M (visibili alla sinistra delle linee di codice in figura) ed il registro coinvolto $xY (nel formato N / M / $xY);
$2)$ le istruzioni tra le quali sono presenti **control hazard** – indicare i numeri di istruzione N ed M (nel formato N / M);
$3)$ quanti cicli di clock sono necessari ad eseguire il programma tramite **forwarding**, spiegando il calcolo effettuato;
$4)$ quanti cicli di clock sarebbero necessari ad eseguire il programma **senza forwarding**, spiegando il calcolo effettuato;
$5)$ quali sono, per ognuna delle cinque fasi, le **istruzioni** (o le bolle) in pipeline durante **il 14° ciclo di clock** (con **forwarding**);
$6)$ Si descriva nel maggior dettaglio possibile il funzionamento dei segnali di controllo ForwardA e ForwardB nella fase EXE del diagramma seguente

![Esame](https://github.com/Elia-Belli/ae2223/assets/107944476/1cac0418-ce14-4ed4-a27e-5cb02ccbd858)
![Architettura Pipeline](https://github.com/Elia-Belli/ae2223/assets/107944476/9527e913-48ac-4c72-bf02-e9015577296e)
---
### Soluzione del Prof
Le soluzioni fornite dal professore potrebbero essere incomplete o poco chiare, quindi sono benvenute soluzioni che le completano, oppure soluzioni alternative e/o con spiegazioni più approfondite: [Soluzione.pdf](https://github.com/sapienzastudentsnetwork/archelab2223/files/11665977/AE.2021-22.-.Appello.S1.-.2022-11-03.-.Esercizio.pipeline.-.Scheda.di.calcolo.pdf)

**Comment by Frisbon** [comment]
In PDF, cosí si legge meglio

[Easy (né vero).pdf](https://github.com/sapienzastudentsnetwork/archelab2223/files/11868299/Easy.ne.vero.pdf)

**Reply by matypist** [reply]
Ottima proposta di soluzione, soltanto una piccola nota (per i posteri che la leggeranno): nel punto **n. 1** dovrebbe essere indicato un **data hazard** anche tra le istruzioni **10** e **12**, con registro coinvolto `$a1`
